apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs

```diff
@@ -0,0 +1,311 @@
+        private static readonly ConcurrentQueue<TaskCompletionSource<int>> s_waitQueue = new ConcurrentQueue<TaskCompletionSource<int>>();
+        private static readonly object s_waitSync = new object();
+
+        private static ConcurrentDictionary<SessionPoolKey, SessionPoolImpl> GlobalPool { get; } =
+            new ConcurrentDictionary<SessionPoolKey, SessionPoolImpl>();
+        private static readonly object s_priorityListSync = new object();
+        private static SortedList<SessionPoolImpl, SessionPoolImpl> PriorityList { get; }
+            = new SortedList<SessionPoolImpl, SessionPoolImpl>();
+
+        static SessionPool()
+        {
+#if NET45
+            AppDomain.CurrentDomain.DomainUnload += (sender, args) =>
+            {
+                Task.Run(ReleaseAll).Wait(s_shutDownTimer);
+            };
+#endif
+#if NETSTANDARD1_5
+            System.Runtime.Loader.AssemblyLoadContext.Default.Unloading += context =>
+            {
+                Task.Run(ReleaseAll).Wait(s_shutDownTimer);
+            };
+#endif
+        }
+
+        private static Task WhenCanceled(this CancellationToken cancellationToken)
+        {
+            var tcs = new TaskCompletionSource<bool>();
+            cancellationToken.Register(s => ((TaskCompletionSource<bool>)s).SetResult(true), tcs);
+            return tcs.Task;
+        }
+
+        private static async Task StartSessionCreating(CancellationToken cancellationToken)
+        {
+            lock (s_waitSync)
+            {
+                if (CurrentActiveSessions < MaximumActiveSessions)
+                {
+                    s_sessionsCreating++;
+                    return;
+                }
+            }
+            var canceledTask = cancellationToken.WhenCanceled();
+            //we need to block or throw because we are out of allocations.
+            while (true)
+            {
+                if (WaitOnResourcesExhausted)
+                {
+                    TaskCompletionSource<int> signal = new TaskCompletionSource<int>();
+                    s_waitQueue.Enqueue(signal);
+                    await Task.WhenAny(signal.Task, canceledTask).ConfigureAwait(false);
+                    cancellationToken.ThrowIfCancellationRequested();
+                    lock (s_waitSync)
+                    {
+                        if (CurrentActiveSessions < MaximumActiveSessions)
+                        {
+                            s_sessionsCreating++;
+                            return;
+                        }
+                    }
+                }
+                else
+                {
+                    throw new RpcException(new Status(StatusCode.ResourceExhausted, "Number of available Sessions exhausted."));
+                }
+            }
+        }
+
+        private static void EndSessionCreating()
+        {
+            lock (s_waitSync)
+            {
+                s_sessionsCreating--;
+            }
+            SignalAnyWaitingRequests();
```

Comment:
i've touched it up a bit.  the important bit I did was allocating the session as soon as the entry is dequeued and task is signaled.  This removes any race conditions and allows the caller who is waiting to know that as soon as his task is signaled, he was already allocated the resource he needs.